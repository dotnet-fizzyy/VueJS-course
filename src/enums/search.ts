export enum SearchByOptionNames {
    Title = 'title',
    Genre = 'genres',
}

export enum SortByOptionsNames {
    ReleaseDate = 'release_date',
    Rating = 'rating',
}

export enum SearchQueryParams {
    SearchTerm = 'term',
    SearchByOption = 'searchBy',
    SortByOption = 'sortBy',
}
