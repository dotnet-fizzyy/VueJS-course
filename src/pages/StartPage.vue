<template>
    <div :class="$style.root">
        <div :class="$style.header">
            <div :class="$style['header-background']">
                <div :class="$style['search-container']">
                    <div :class="$style['logo-container']">
                        <logo-icon />
                    </div>

                    <template v-if="!!selectedFilm">
                        <div :class="$style['search-icon-container']" @click="onBackToSearch">
                            <search-icon />
                        </div>

                        <film-full-description
                            :id="selectedFilm.id"
                            :name="selectedFilm.name"
                            :short-description="selectedFilm.shortDescription"
                            :full-description="selectedFilm.fullDescription"
                            :release-year="selectedFilm.releaseYear"
                            :poster-url="selectedFilm.posterUrl"
                            :poster-alt="selectedFilm.posterAlt"
                            :rating="selectedFilm.rating"
                            :runtime="selectedFilm.runtime"
                        />
                    </template>
                    <template v-else>
                        <search-panel />
                    </template>
                </div>
            </div>
        </div>

        <div :class="$style.body">
            <div :class="$style['sort-panel-container']">
                <sort-panel :available-items-count="7" />
            </div>

            <div :class="$style['films-display']">
                <div v-for="film in films" :key="film.id" :class="$style['film-container']">
                    <film-short-description
                        :id="film.id"
                        :name="film.name"
                        :short-description="film.shortDescription"
                        :release-year="film.releaseYear"
                        :poster-url="film.posterUrl"
                        :poster-alt="film.posterAlt"
                        @select="selectFilm"
                    />
                </div>
            </div>
        </div>

        <div :class="$style.footer">
            <logo-icon />
        </div>
    </div>
</template>

<script lang="ts">
import FilmFullDescription from '@/components/films/FilmFullDescription.vue';
import FilmShortDescription from '@/components/films/FilmPreview.vue';
import LogoIcon from '@/components/icons/LogoIcon.vue';
import SearchIcon from '@/components/icons/SearchIcon.vue';
import SearchPanel from '@/components/search/SearchPanel.vue';
import SortPanel from '@/components/listing/SortPanel.vue';
import Vue from 'vue';
import { fullFilmsDescription, shortFilmsDescription } from '@/mocks/mockFilms';

export default Vue.extend({
    name: 'StartPage',
    components: { LogoIcon, SearchIcon, FilmFullDescription, FilmShortDescription, SearchPanel, SortPanel },
    data() {
        return {
            films: shortFilmsDescription,
            selectedFilm: null,
        };
    },
    methods: {
        selectFilm(id: string): void {
            this.selectedFilm = fullFilmsDescription.find(x => x.id === id);
        },
        onBackToSearch(): void {
            this.selectedFilm = null;
        },
    },
});
</script>

<style lang="scss" module scoped>
@import '../styles/fonts';

.root {
    width: 100%;
    height: 100%;
    display: flex;
    flex-direction: column;
}

.header {
    background-image: url('https://res.cloudinary.com/djlynoeio/image/upload/v1649880735/vue/film-posters.jpg');
    background-position: center;
    background-repeat: no-repeat;
    background-size: cover;
}

.header-background {
    height: 100%;
    background-color: getRgbaBlack(0.8);
    backdrop-filter: blur(3px);
    position: relative;
}

.logo-container {
    position: absolute;
    left: 60px;
    top: 20px;
}

.search-container {
    width: 100%;
    height: 100%;
    display: flex;
    align-items: center;
    justify-content: space-between;
    box-sizing: border-box;
    padding: 100px 120px;
}

.search-icon-container {
    position: absolute;
    right: 60px;
    top: 20px;
    cursor: pointer;
}

.body {
}

.films-display {
    display: flex;
    flex-direction: row;
    flex-wrap: wrap;
    padding: 40px;
}

.film-container {
    padding: 25px;
}

.sort-panel-container {
    height: 70px;
    background-color: $light-grey;
    display: flex;
    align-items: center;
    flex-direction: row;
    padding: 0 120px;
}

.footer {
    height: 70px;
    background-color: $light-grey;
    margin-top: auto;
    display: flex;
    align-items: center;
    justify-content: center;
}
</style>
