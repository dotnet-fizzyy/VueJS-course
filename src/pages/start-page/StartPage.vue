<template>
    <div :class="$style.root">
        <start-page-header :selected-film-id="selectedFilmId" @back-to-search="backToSearch" />

        <start-page-body :select-film="selectFilm" />

        <start-page-footer />
    </div>
</template>

<script lang="ts">
import StartPageBody from '@/pages/start-page/StartPageBody.vue';
import StartPageFooter from '@/pages/start-page/StartPageFooter.vue';
import StartPageHeader from '@/pages/start-page/StartPageHeader.vue';
import Vue from 'vue';

export default Vue.extend({
    name: 'StartPage',
    components: {
        StartPageFooter,
        StartPageHeader,
        StartPageBody,
    },
    data() {
        return {
            selectedFilmId: '',
        };
    },
    methods: {
        selectFilm(id: string): void {
            this.selectedFilmId = id;
        },
        backToSearch(): void {
            this.selectedFilmId = '';
        },
    },
});
</script>

<style lang="scss" module scoped>
.root {
    width: 100%;
    height: 100%;
    display: flex;
    flex-direction: column;
}
</style>
