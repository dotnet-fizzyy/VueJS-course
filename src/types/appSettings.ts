export interface AppSettings {
    isDevelopmentMode: boolean;
    defaultPosterUrl: string;
    defaultPosterAlt: string;
    apiUrl: string;
}
