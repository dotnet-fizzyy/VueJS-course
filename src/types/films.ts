export interface Film {
    id: string;
    name: string;
    releaseYear: number;
    posterUrl?: string;
    posterAlt?: string;
    shortDescription: string;
}

export interface FilmShortDescription extends Film {}

export interface FilmFullDescription extends Film {
    rating: number;
    runtime: number;
    fullDescription: string;
}
