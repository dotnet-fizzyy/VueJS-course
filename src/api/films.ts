import Vue from 'vue';
import VueResource from 'vue-resource';
import { CollectionResponse, Film, FilmResponse } from '@/types/films';
import { mapFilmApiResponseToFilm } from '@/utils/films';

export default class FilmsApi {
    private static readonly Headers: Record<string, string> = {
        Accept: 'application/json',
        'Content-Type': 'application/json',
    };

    public static async getFilms(): Promise<CollectionResponse<Film>> {
        const response: VueResource.HttpResponse = await Vue.http.get(`${FilmsApi.getBaseUrl()}/movies`, {
            headers: FilmsApi.Headers,
        });
        const responseBody: CollectionResponse<FilmResponse> = await response.json();

        return {
            total: responseBody.total,
            limit: responseBody.limit,
            offset: responseBody.offset,
            data: responseBody.data.map(mapFilmApiResponseToFilm),
        } as CollectionResponse<Film>;
    }

    public static async getFilmById(id: string): Promise<Film> {
        const response: VueResource.HttpResponse = await Vue.http.get(`${FilmsApi.getBaseUrl()}/movies/${id}`, {
            headers: FilmsApi.Headers,
        });
        const responseBody: FilmResponse = await response.json();

        return mapFilmApiResponseToFilm(responseBody) as Film;
    }

    private static getBaseUrl(): string {
        const baseUrl = Vue.http?.options?.root;

        if (!baseUrl) {
            throw new Error('Base URL for communicating with API is not defined');
        }

        return baseUrl;
    }
}
