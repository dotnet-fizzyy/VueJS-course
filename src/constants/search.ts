import { ButtonGroupOption } from '@/types/button';
import { SearchByOptionNames, SortByOptionsNames } from '@/enums/search';

export const SortByOptions: ButtonGroupOption[] = [
    { label: 'Release Date', name: SortByOptionsNames.ReleaseDate },
    { label: 'Rating', name: SortByOptionsNames.Rating },
];

export const SearchByOptions: ButtonGroupOption[] = [
    { label: 'Title', name: SearchByOptionNames.Title },
    { label: 'Genre', name: SearchByOptionNames.Genre },
];
