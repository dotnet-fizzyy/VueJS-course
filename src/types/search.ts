import { SearchQueryParams } from '@/enums/search';

export interface SearchParams {
    [SearchQueryParams.SearchTerm]?: string;
    [SearchQueryParams.SearchByOption]: string;
    [SearchQueryParams.SortByOption]: string;
}
