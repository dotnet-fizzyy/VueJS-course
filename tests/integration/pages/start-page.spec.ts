import Films from '@/vuex/modules/films/state';
import FiltersPlugin from '@/plugins/filtersPlugin';
import MockRoutes from '@/mocks/mockEndpoints';
import StartPage from '@/pages/start-page/StartPage.vue';
import Vue from 'vue';
import VueLazyload from 'vue-lazyload';
import VueResource from 'vue-resource';
import Vuex from 'vuex';
import { FilmGetterProps } from '@/vuex/modules/films/getters';
import { FilmsModuleName } from '@/vuex/store/constants';
import { NoFilmsFoundMessage } from '@/constants/search';
import { State } from '@/vuex/store/state';
import { createLocalVue, mount } from '@vue/test-utils';
import { delay } from '@/utils';
import { getAppSettings } from '@/utils/appSettings';

const localVue = createLocalVue();
localVue.use(VueResource);
localVue.use(Vuex);
localVue.use(FiltersPlugin);
localVue.use(VueLazyload);

localVue.http.options.root = getAppSettings().apiUrl;

// todo: try to find the other way to set "http" to global Vue
Vue.http = localVue.http;

localVue.http.interceptors.push(((request, next) => {
    const route = MockRoutes.find(item => {
        return request.method === item.method && request.url === item.url;
    });

    if (route) {
        next(request.respondWith(JSON.stringify(route.response), { status: 200 }));
    } else {
        next(request.respondWith({ status: 404, statusText: 'Oh no! Not found!' }));
    }
}) as VueResource.HttpInterceptor);

describe('StartPage.vue', () => {
    test('Should render', () => {
        //Arrange
        const store = new Vuex.Store({
            modules: { [FilmsModuleName]: Films },
        });

        const wrapper = mount(StartPage, {
            store,
            localVue,
        });

        //Act & Assert
        expect(wrapper.element).toMatchSnapshot();
    });

    test('Should display all available films previews in body section from store', () => {
        //Arrange
        const store = new Vuex.Store({
            modules: { [FilmsModuleName]: Films },
        });

        const wrapper = mount(StartPage, {
            store,
            localVue,
        });

        const expectedItemsCount: number = (store.state as State).films.length;

        //Act
        const filmPreviewsWrappers = wrapper.findAll('[data-aqa-preview]');

        //Assert
        expect(filmPreviewsWrappers.length).toEqual(expectedItemsCount);
    });

    test(`Should display "${NoFilmsFoundMessage}" label if film previews empty array was provided from store`, () => {
        //Arrange
        const store = new Vuex.Store({
            modules: {
                [FilmsModuleName]: {
                    ...Films,
                    getters: {
                        ...Films.getters,
                        [FilmGetterProps.FilmsPreviews]: () => [],
                        [FilmGetterProps.IsLoadingItems]: () => false,
                    },
                },
            },
        });

        const wrapper = mount(StartPage, {
            store,
            localVue,
        });

        //Act & Assert
        const noFilmsMessageContainerWrapper = wrapper.find('#no-films-message-container');

        expect(noFilmsMessageContainerWrapper.element).toBeTruthy();

        const noFilmsMessageWrapper = noFilmsMessageContainerWrapper.find('span');

        expect(noFilmsMessageWrapper.text()).toEqual(NoFilmsFoundMessage);
    });

    test(`Should display film full description if it was selected from previews`, async () => {
        //Arrange
        const requestDelay: number = 100;

        const store = new Vuex.Store({
            modules: { [FilmsModuleName]: Films },
        });

        const wrapper = mount(StartPage, {
            store,
            localVue,
        });

        //Act & Assert
        await delay(requestDelay);
        await wrapper.vm.$nextTick();

        const filmPreviewWrapper = wrapper.find('[data-aqa-preview]');
        expect(filmPreviewWrapper.element).toBeTruthy();

        const filmId = filmPreviewWrapper.props('id');
        filmPreviewWrapper.vm.$emit('select', filmId);

        expect(filmPreviewWrapper.emitted('select')).toBeTruthy();

        await delay(requestDelay);
        await wrapper.vm.$nextTick();

        const filmFullDescriptionWrapper = wrapper.find('[data-aqa-full-description]');

        expect(filmFullDescriptionWrapper.element).toBeTruthy();

        const filmPreviewNameWrapper = filmPreviewWrapper.find('[data-aqa-film-preview-name]');
        const filmFullDescriptionNameWrapper = filmFullDescriptionWrapper.find('[data-aqa-full-desc-name]');

        expect(filmPreviewNameWrapper.element).toBeTruthy();
        expect(filmFullDescriptionNameWrapper.element).toBeTruthy();
        expect(filmPreviewNameWrapper.text()).toEqual(filmFullDescriptionNameWrapper.text());
    });
});
