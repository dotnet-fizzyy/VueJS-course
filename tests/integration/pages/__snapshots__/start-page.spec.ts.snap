// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StartPage.vue Should render 1`] = `
<div
  class="root"
>
  <div
    class="root"
    data-aqa-start-header=""
  >
    <div
      class="background"
    >
      <div
        class="logo-icon-container"
      >
        <div>
          <span
            class="first-logo-part"
          >
            Netflix
          </span>
           
          <span
            class="second-logo-part"
          >
            Roulette
          </span>
        </div>
      </div>
       
      <div
        class="root"
        data-aqa-search-panel=""
      >
        <span
          class="main-label"
        >
          Find your movie
        </span>
         
        <div
          class="search-container"
        >
          <div
            class="input-container"
          >
            <div
              class="root"
            >
              <!---->
               
              <input
                class="input"
                placeholder="Search"
                style="font-size: 24px;"
              />
               
              <!---->
            </div>
          </div>
           
          <div
            class="search-button-container"
          >
            <button
              class="primary-button normal"
              style="font-size: 24px;"
            >
              Search
            </button>
          </div>
        </div>
         
        <div
          class="options-container"
        >
          <div
            class="root"
          >
            <span
              class="label"
            >
              Search By
            </span>
             
            <div
              class="options-container"
            >
              <button
                class="selected-option option"
                id="title"
              >
                
                Title
            
              </button>
              <button
                class="non-selected-option option"
                id="genres"
              >
                
                Genre
            
              </button>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
   
  <div
    data-aqa-start-body=""
  >
    <div
      class="sort-panel-container"
    >
      <div
        class="root"
      >
        <span
          class="title"
          id="available-items-count"
        >
          0 movies found
        </span>
         
        <div
          class="root"
        >
          <span
            class="label"
          >
            Sort By
          </span>
           
          <div
            class="options-container"
          >
            <button
              class="selected-option option"
              id="release_date"
            >
              
                Release Date
            
            </button>
            <button
              class="non-selected-option option"
              id="rating"
            >
              
                Rating
            
            </button>
          </div>
        </div>
      </div>
    </div>
     
    <div
      class="no-films-found-container"
      id="no-films-message-container"
    >
      <svg
        class="root"
        viewBox="0 0 512 512"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M304 48C304 74.51 282.5 96 256 96C229.5 96 208 74.51 208 48C208 21.49 229.5 0 256 0C282.5 0 304 21.49 304 48zM304 464C304 490.5 282.5 512 256 512C229.5 512 208 490.5 208 464C208 437.5 229.5 416 256 416C282.5 416 304 437.5 304 464zM0 256C0 229.5 21.49 208 48 208C74.51 208 96 229.5 96 256C96 282.5 74.51 304 48 304C21.49 304 0 282.5 0 256zM512 256C512 282.5 490.5 304 464 304C437.5 304 416 282.5 416 256C416 229.5 437.5 208 464 208C490.5 208 512 229.5 512 256zM74.98 437C56.23 418.3 56.23 387.9 74.98 369.1C93.73 350.4 124.1 350.4 142.9 369.1C161.6 387.9 161.6 418.3 142.9 437C124.1 455.8 93.73 455.8 74.98 437V437zM142.9 142.9C124.1 161.6 93.73 161.6 74.98 142.9C56.24 124.1 56.24 93.73 74.98 74.98C93.73 56.23 124.1 56.23 142.9 74.98C161.6 93.73 161.6 124.1 142.9 142.9zM369.1 369.1C387.9 350.4 418.3 350.4 437 369.1C455.8 387.9 455.8 418.3 437 437C418.3 455.8 387.9 455.8 369.1 437C350.4 418.3 350.4 387.9 369.1 369.1V369.1z"
        />
      </svg>
    </div>
  </div>
   
  <div
    class="root"
    data-aqa-start-footer=""
  >
    <div>
      <span
        class="first-logo-part"
      >
        Netflix
      </span>
       
      <span
        class="second-logo-part"
      >
        Roulette
      </span>
    </div>
  </div>
</div>
`;
